records/2024-04-11.2FA-magic-link.md

```diff
@@ -0,0 +1,171 @@
+
+1. Access to a user's partially authenticated cookie.
+2. Access to  the user's email inbox to obtain their magic link.
+   * Man-in-the-middle attack to obtain the user's magic link.
+
+
+#### Mitigations
+* CSRF protections buit into Flask, and which are already enabled in our application, configure the `Set-Cookie` header with the `HttpOnly` and `Secure` options, preventing Javascript from reading the contents of the session cookie, and limits the cookie(s) to HTTPS traffic only
+
+* The `Strict-Transport-Security` header is enabled site wide, converting all requests to HTTPS, thus mitigating man-in-the-middle attacks.
+
+* Introduce a process for checking that the `Origin` header matches the target origin when a magic link is followed. As our application is distributed and lives behind a load balancer, making each node aware of it's target origin could prove to be non-trivial task and/or have unintended consequences.
+
+See the [Fallback Mitigation section](#Fallback-Mitigation---An-over-engineered-"last-resort"-solution)
+
+### Email Phishing
+An attacker could tamper with the 2FA email we send to the user and forward them to a phishing site that prompts them to re-input their username and password. If the user is not aware of this type of risk they could unintentionally provide an attacker with their log in credentials.
+
+In this instance, the attacker now has a means to create their own partially authenticated session cookie, and would only need access to the magic link. If they have successfully tampered with the 2FA email, we can assume the attacker already has the means to obtain the magic link, even without direct access to the user's inbox.
+
+#### Mitigations
+* Provide guidance to inform users of the risks of magic links, and steps they can take to mitigate them.
+* See the [Fallback Mitigation section](#Fallback-Mitigation---An-over-engineered-"last-resort"-solution)
+
+###
+
+## Fallback Mitigation - An over-engineered "last resort" solution
+
+A fallback mitigation layer to protect against scenarios that are out of our control would be to establish a process for generating a unique visitor fingerprint. This fingerprint could be ephemerally stored for the duration of a magic link's validity period. That fingerprint could then be validated once the user clicks the magic link, and before beginning the final steps of the login process.
+
+The fingerprint could use the user's browser attributes, IP address, HTTPS session keys, etc to form a composite key that could then be hashed to uniquely identify the user. Since the success of the attacker is time dependent, [Argon2d](https://github.com/P-H-C/phc-winner-argon2) could be a good candidate for such a hashing implementation as it uses data-depending memory access to make it resistant to GPU based brute force attacks.
```

Comment:
Did you speak with Andrew about using the `secrets` module too (instead of `SystemRandom`)? Do you know how that would compare?

Also, do we consider lowering the time period to use the link? At the moment, the default is 30 minutes for the code sent via the email.

Reply:
> Did you speak with Andrew about using the secrets module too (instead of SystemRandom)? Do you know how that would compare?

I have not, thank you for the reminder about the `secrets` module. I had forgotten about it but I think it could be applicable here - will follow up with Andrew.

> do we consider lowering the time period to use the link? At the moment, the default is 30 minutes for the code sent via the email.

30 minutes is quite lenient / lax to me, I think reducing this time is a reasonable mitigation that we can consider. I will add a section about this.

Reply:
Added the sections addressing the `secrets` module migration and detailing reduction of the 2FA validity period as a potential mitigation.